cmd/launcher/launcher.go

```diff
@@ -235,25 +176,143 @@ func runLauncher(ctx context.Context, cancel func(), opts *launcher.Options) err
+		}
+	}
+	level.Info(logger).Log(
+		"msg", "using default system root directory",
+		"path", rootDirectory,
+	)
+
+	if err := os.MkdirAll(rootDirectory, 0700); err != nil {
+		return "", fmt.Errorf("creating root directory: %w", err)
+	}
+
+	return rootDirectory, nil
+}
+
+func initHttpClient(opts *launcher.Options) *http.Client {
+	// construct the appropriate http client based on security settings
+	httpClient := http.DefaultClient
+	if opts.InsecureTLS {
+		httpClient = &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: &tls.Config{
+					InsecureSkipVerify: true,
+				},
+			},
+		}
+	}
+	return httpClient
+}
+
+func initBoltDb(dir string) (*bbolt.DB, error) {
+	boltOptions := &bbolt.Options{Timeout: time.Duration(30) * time.Second}
+	db, err := bbolt.Open(filepath.Join(dir, "launcher.db"), 0600, boltOptions)
+	if err != nil {
+		return nil, fmt.Errorf("opening launcher db: %w", err)
+	}
+	return db, nil
+}
+
+func createCertificatePool(opts *launcher.Options) (*x509.CertPool, error) {
+	// create the certificate pool
+	var rootPool *x509.CertPool
+	if opts.RootPEM != "" {
+		rootPool = x509.NewCertPool()
+		pemContents, err := ioutil.ReadFile(opts.RootPEM)
+		if err != nil {
+			return nil, fmt.Errorf("reading root certs PEM at path %s: %w", opts.RootPEM, err)
+		}
+		if ok := rootPool.AppendCertsFromPEM(pemContents); !ok {
+			return nil, fmt.Errorf("appending root certs PEM at path %s: %w", opts.RootPEM, err)
+		}
+	}
+	return rootPool, nil
+}
+
+func handleSignals(logger log.Logger, cancel func(), runGroup *run.Group) chan os.Signal {
+	// Create a channel for signals
+	sigChannel := make(chan os.Signal, 1)
+
+	// Add a rungroup to catch things on the sigChannel
+	// Attach a notifier for os.Interrupt
+	runGroup.Add(func() error {
+		signal.Notify(sigChannel, os.Interrupt)
+		select {
+		case <-sigChannel:
+			level.Info(logger).Log("msg", "beginnning shutdown via signal")
+			return nil
+		}
+	}, func(err error) {
+		level.Info(logger).Log("msg", "interrupted", "err", err)
+		level.Debug(logger).Log("msg", "interrupted", "err", err, "stack", fmt.Sprintf("%+v", err))
+		cancel()
+		close(sigChannel)
+	})
+
+	return sigChannel
+}
+
+func createSystrayRunner(logger log.Logger, runGroup *run.Group, opts *launcher.Options) *systrayruntime.SystrayUsersProcessesRunner {
 	if (opts.KolideServerURL == "k2device-preprod.kolide.com" || opts.KolideServerURL == "localhost:3443") && runtime.GOOS == "darwin" {
 		systrayRunner := systrayruntime.New(logger, time.Second*5)
 		runGroup.Add(systrayRunner.Execute, systrayRunner.Interrupt)
+		return systrayRunner
 	}
 
-	err = runGroup.Run()
-	return errors.Wrap(err, "run service")
+	return nil
 }
 
-func writePidFile(path string) error {
-	err := ioutil.WriteFile(path, []byte(strconv.Itoa(os.Getpid())), 0600)
-	return errors.Wrap(err, "writing pidfile")
+func setupControlServer(ctx context.Context, logger log.Logger, db *bbolt.DB, runGroup *run.Group, opts *launcher.Options) error {
```

Comment:
Picking this place to comment, I feel weird about passing in runGroup. Talking aloud, `runLauncher is basically a main function. And I think of main functions as being about wiring together pieces. From that perspective something like:
```
controlServer, err := setupControlServer(...)
if err != nil { ... }
runGroup.Add(...)
```

doesn't feel wrong, but also feels kinda needless verbose, and creates that superfluous `controlServer`.

So I wonder about a helper to add things to the rungroup. Using a simplified example, something like https://go.dev/play/p/wLX_2qkaLsg

```
func main() {

	if err := rgHelper(newBroke()); err != nil {
		fmt.Println("ERR:", err)
	}

	if err := rgHelper(newFoo()); err != nil {
		fmt.Println("ERR:", err)
	}

}

func rungroupAdd(one, two string) {
	fmt.Printf("one: %s, two: %s\n", one, two)
}

func newFoo() (string, string, error) {
	return "foo", "bar", nil
}

func newBroke() (string, string, error) {
	return "", "", errors.New("broke")
}

func rgHelper(one, two string, err error) error {
	if err != nil {
		return err
	}

	rungroupAdd(one, two)
	return nil
}
```

Can be cleaned up a little with an interface around Execute and Interrupt